File: python/5.py
with open("../5.txt") as f:
    ranges: list[range] = []
    ingredients = []
    for line in f:
        line = line.strip()
        if "-" not in line:
            ingredients.append(int(line))
            break
        start, end = [int(x) for x in line.split("-")]
        whole = False
        kill: list[range] = []
        while True:
            new_range = range(start, end + 1)
            for i, r in enumerate(ranges):
                if r.start in new_range and r.stop - 1 in new_range:
                    kill.append(r)
                    continue
                if start in r:
                    if end in r:
                        whole = True
                        continue
                    start = r.start
                    break
                if end in r:
                    end = r.stop - 1
                    break
            else:
                break
            ranges.pop(i)
        if whole:
            continue
        ranges.append(new_range)
        for k in kill:
            ranges.remove(k)
    for line in f:
        ingredients.append(int(line))

print(sum(1 for i in ingredients if any(i in r for r in ranges)))

print(sum(r.stop - r.start for r in ranges))

File: rust/src/bin/5.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::ops::RangeInclusive;
use std::str::FromStr;

fn main() {
    let mut lines = BufReader::new(File::open("../5.txt").unwrap())
        .lines()
        .map(|l| l.unwrap());
    let mut ranges: Vec<RangeInclusive<usize>> = Vec::new();
    let mut ingredients = Vec::new();
    'outer: for line in lines.by_ref() {
        if !line.contains('-') {
            ingredients.push(usize::from_str(&line).unwrap());
            break;
        }
        let (start, end) = line.split_once('-').unwrap();
        let mut start = usize::from_str(start).unwrap();
        let mut end = usize::from_str(end).unwrap();
        let mut kill = Vec::new();
        let mut new_range;
        'inner: loop {
            new_range = start..=end;
            for (i, r) in ranges.iter().enumerate() {
                if new_range.contains(r.start()) && new_range.contains(r.end()) {
                    kill.push(r.clone());
                    continue;
                }
                if r.contains(&start) {
                    if r.contains(&end) {
                        continue 'outer;
                    }
                    start = *r.start();
                    ranges.swap_remove(i);
                    continue 'inner;
                }
                if r.contains(&end) {
                    end = *r.end();
                    ranges.swap_remove(i);
                    continue 'inner;
                }
            }
            break;
        }
        ranges.push(new_range);
        for k in kill {
            ranges.swap_remove(ranges.iter().position(|p| *p == k).unwrap());
        }
    }
    for line in lines {
        ingredients.push(usize::from_str(&line).unwrap());
    }

    println!(
        "{}",
        ingredients
            .iter()
            .filter(|i| ranges.iter().any(|r| r.contains(i)))
            .count()
    );

    println!(
        "{}",
        ranges
            .iter()
            .map(|r| r.end() - r.start() + 1)
            .sum::<usize>()
    );
}
